web/source/osk/banner.ts

```diff
@@ -0,0 +1,259 @@
+
+      if (this.div) {
+        let ds = this.div.style;
+        ds.display=(this._Visible) ? 'block': 'none';
+      }
+    }
+
+    public get enable():boolean {
+      return this._Enabled;
+    }
+
+    public set enable(enable: boolean) {
+      this._Enabled = enable;
+    }
+
+    public constructor(visible: boolean, enabled: boolean, height?: number) {
+      let keymanweb = com.keyman.singleton;
+      let util = keymanweb.util;
+
+      let d = util._CreateElement('div');
+      d.id = "keymanweb_banner_bar";
+      d.className = "kmw-banner-bar";
+      this.div = d;
+
+      this.height = (height >= 0) ? height : this.DEFAULT_HEIGHT;
+      this.visible = visible;
+      this.enable = enabled;
+    }
+
+    public appendStyleSheet() {
+      let keymanweb = com.keyman.singleton;
+      let util = keymanweb.util;
+
+      // TODO: add stylesheets
+    }
+
+    public getDiv():HTMLElement {
+      return this.div;
+    }
+  }
+
+  /**
+   * Function       BlankBanner
+   * Description    A banner of height 0 that should not be shown
+   */
+  export class BlankBanner extends Banner {
+
+    constructor() {
+      super(false, true, 0);
+    }
+  }
+
+  /**
+   * Function       ImageBanner
+   * @param         {string}        imagePath   Path of image to display in the banner
+   * Description    Display an image in the banner
+   */
+  export class ImageBanner extends Banner {
+    private img: HTMLElement;
+
+    constructor(imagePath, height?: number) {
+      if (imagePath.length > 0) {
+        super(true, true);
+      } else {
+        super(false, true, 0);
+      }
+
+      this.img = document.createElement('img');
+      this.img.setAttribute('src', imagePath);
+      let ds = this.img.style;
+      ds.width = '100%';
+      ds.height = this.height + 'px';
+      if (this.div) {
+        this.div.appendChild(this.img);
+      }
+    }
+
+    /**
+     * Function     setImagePath
+     * Scope        Public
+     * @param       {string}     imagePath   Path of image to display in the banner
+     * Description  Update the image in the banner
+     */
+    public setImagePath(imagePath: string) {
+      if (this.img) {
+        this.img.setAttribute('src', imagePath);
+      }
+      if (imagePath.length > 0) {
+        this.height = this.DEFAULT_HEIGHT;
+        this.visible = true;
+        this.enable = true;
+
+        if (this.img) {
+          let ds = this.img.style;
+          ds.height = this.height + 'px';
+        }
+      }
+    }
+  }
+
+  export class BannerSuggestionSpec {
+    id: string;
+    languageID: string;
+    text?: string;
+    width: string;
+    pad?: string;
+    widthpc?: number; // Added during OSK construction.
+    padpc?: number; // Added during OSK construction.
+
+    constructor(id: string, languageID: string, text?: string, width?: string, pad?: string) {
+      this.id = id;
+      this.languageID = languageID;
+      this.text = text;
+      this.width = width ? width : "50";
```

Comment:
Why is width a number in quotes? Is it pixels? I also wonder why the field needs a width specified?

Reply:
This was based on visualKeyboard.OSKKeySpec where width is a string. The `width` field will later be used to check the text proportion

Reply:
Okay, so this is a nasty that I don't think we should propagate into new code. Can we eliminate the extra width property and have a single numeric property?